slsim/Sources/QuasarCatalog/quasar_pop.py

```diff
@@ -0,0 +1,355 @@
+            converted_magnitude = magnitude + DM + K_corr
+        else:
+            raise ValueError(
+                "Conversion must be either 'apparent_to_absolute' or 'absolute_to_apparent'"
+            )
+
+        return converted_magnitude
+
+    def n_comoving(self, m_min, m_max, z_value):
+        """Calculates the comoving number density of quasars for a given redshift by
+        integrating dPhi/dM over the range of apparent magnitudes.
+
+        :param m_min: Minimum apparent magnitude.
+        :type m_min: float or np.ndarray
+        :param m_max: Maximum apparent magnitude.
+        :type m_max: float or np.ndarray
+        :param z_value: Redshift value.
+        :type z_value: float or np.ndarray
+        :return: Comoving number density of quasars.
+        :rtype: float or np.ndarray :unit: Mpc^-3
+        """
+        M_min = self.convert_magnitude(
+            m_min, z_value, conversion="apparent_to_absolute"
+        )
+        M_max = self.convert_magnitude(
+            m_max, z_value, conversion="apparent_to_absolute"
+        )
+
+        def integrand(M, z):
+            return self.dPhi_dM(M, z)
+
+        # If z_value is an array, apply integration for each element
+        if isinstance(z_value, np.ndarray):
+            integrals = []
+            for z in z_value:
+                integral, _ = quad(integrand, M_min, M_max, args=(z,))
+                if integral < 0 or np.isnan(integral):
+                    print(f"Invalid integral value: {integral} for z = {z}")
+                integrals.append(integral)
+            return np.array(integrals)
+        else:
+            integral, _ = quad(integrand, M_min, M_max, args=(z_value,))
+            if integral < 0 or np.isnan(integral):
+                print(f"Invalid integral value: {integral} for z = {z_value}")
+            return integral
+
+    def generate_quasar_redshifts(self, m_min, m_max, return_single=False):
+        """Generates redshift locations of quasars using a light cone formulation.
+
+        :param m_min: Minimum apparent magnitude.
+        :type m_min: float
+        :param m_max: Maximum apparent magnitude.
+        :type m_max: float
+        :param return_single: If True, return a single redshift value instead of an
+            array.
+        :type return_single: bool
+        :return: Redshift locations of quasars.
+        :rtype: float or np.ndarray
+        """
+        n_comoving_values = np.array(
+            [self.n_comoving(m_min, m_max, z) for z in self.redshifts]
+        )
+
+        sampled_redshifts = redshifts_from_comoving_density(
+            redshift=self.redshifts,
+            density=n_comoving_values,
+            sky_area=self.sky_area,
+            cosmology=self.cosmo,
+            noise=self.noise,
+        )
+
+        # Ensure redshifts are stored as numpy array
+        sampled_redshifts = np.array(sampled_redshifts, dtype=float)
+
+        if return_single:
```

Comment:
Do we need this option?

Reply:
Changed the code so that a numpy array for the redshift values is always generated, even if a single redshift value is generated.